### PR TITLE
pdh support

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
This is a work in progress. I still need to bind types and move content into um/pdh.

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
I'm having issues with unions. Otherwise I *think* I'm almost done in here.

File: src/um/pdh.rs [hunk 1]
```diff
@@ -54,3 +208,338 @@ STRUCT!{struct PDH_COUNTER_PATH_ELEMENTS_W {
...
+    hQuery: PDH_HQUERY, szFullCounterPath: LPCSTR, dwUserData: DWORD_PTR,
+    phCounter: *mut PDH_HCOUNTER
+) -> PDH_STATUS}
+EXTERN!{stdcall fn PdhCollectQueryDataWithTime(
+    hQuery: PDH_HQUERY, pllTimeStamp: *mut LONGLONG
+) -> PDH_STATUS}
+EXTERN!{stdcall fn PdhValidatePathExW(
+    hDataSource: PDH_HLOG, szFullPathBuffer: LPCWSTR
+) -> PDH_STATUS}
+EXTERN!{stdcall fn PdhValidatePathExA(
```
Comment (docbrown) [inline src/um/pdh.rs]: This and other *A functions take a `LPCSTR`, not a `LPCTSTR`.

Comment (GuillaumeGomez) [inline src/um/pdh.rs]: Hum no: https://msdn.microsoft.com/en-us/library/windows/desktop/aa372992(v=vs.85).aspx

Also, `LPCTSTR` is an alias to `LPCSTR` in non-unicode case.

Comment (yupi2) [inline src/um/pdh.rs]: doc is correct. Function calls that use TCHAR and TSTR varients expand to the *W or *A version of the function at the flip of the unicode switch in code. So it's not good practice to use TSTRs and TCHARs when using the *W or *A version specifically.

File: src/um/pdh.rs [hunk 2]
```diff
@@ -23,17 +28,166 @@ pub type PDH_HQUERY = HANDLE;
...
+/// bSingleCounterPerDialog  :1
+/// bLocalCountersOnly       :1
+/// bWildCardInstances       :1
+/// bHideDetailBox           :1
+/// bInitializePath          :1
+/// bDisableMachineSelection :1
+/// bIncludeCostlyObjects    :1
+/// bShowObjectBrowser       :1
+/// bReserved                :22
+STRUCT!{struct PDH_BROWSE_DLG_CONFIG {
```
Comment (yupi2) [inline src/um/pdh.rs]: Should be PDH_BROWSE_DLG_CONFIG_A and also should have PDH_BROWSE_DLG_CONFIG_W added (along with their A/W versions being used in the correct functions.

File: src/um/pdh.rs [hunk 3]
```diff
@@ -54,3 +208,338 @@ STRUCT!{struct PDH_COUNTER_PATH_ELEMENTS_W {
...
+    hDataSource: PDH_HLOG, szMachineName: LPCTSTR, szObjectName: LPCTSTR,
+    szDefaultCounterName: LPTSTR, pcchBufferSize: LPDWORD
+) -> PDH_STATUS}
+EXTERN!{stdcall fn PdhBrowseCountersHW(
+    pBrowseDlgData: PPDH_BROWSE_DLG_CONFIG
+) -> PDH_STATUS}
+EXTERN!{stdcall fn PdhBrowseCountersHA(
+    pBrowseDlgData: PPDH_BROWSE_DLG_CONFIG
+) -> PDH_STATUS}
+EXTERN!{stdcall fn PdhEnumLogSetNamesW(
```
Comment (yupi2) [inline src/um/pdh.rs]: PdhEnumLogSetNamesW and PdhEnumLogSetNamesA have swapped string types.

File: src/um/pdh.rs [hunk 4]
```diff
@@ -54,3 +208,338 @@ STRUCT!{struct PDH_COUNTER_PATH_ELEMENTS_W {
...
+    szDefaultCounterName: LPWSTR, pcchBufferSize: LPDWORD
+) -> PDH_STATUS}
+EXTERN!{stdcall fn PdhGetDefaultPerfCounterA(
+    szDataSource: LPCTSTR, szMachineName: LPCTSTR, szObjectName: LPCTSTR,
+    szDefaultCounterName: LPTSTR, pcchBufferSize: LPDWORD
+) -> PDH_STATUS}
+EXTERN!{stdcall fn PdhBrowseCountersW(
+    pBrowseDlgData: PPDH_BROWSE_DLG_CONFIG
+) -> PDH_STATUS}
+EXTERN!{stdcall fn PdhBrowseCountersA(
```
Comment (yupi2) [inline src/um/pdh.rs]: PdhBrowseCountersW and PdhBrowseCountersA should use PPDH_BROWSE_DLG_CONFIG_* types.

**Comment by GuillaumeGomez** [comment]
Updated.

File: src/um/pdh.rs (hunk 1 above)
Comment (GuillaumeGomez) [inline src/um/pdh.rs]: Ok, will update it in this way then.

**Comment by GuillaumeGomez** [comment]
I think I'm good?

File: src/um/pdh.rs [hunk 5]
```diff
@@ -54,3 +222,338 @@ STRUCT!{struct PDH_COUNTER_PATH_ELEMENTS_W {
...
+EXTERN!{stdcall fn PdhMakeCounterPathA(
+    pCounterPathElements: PPDH_COUNTER_PATH_ELEMENTS_A, szFullPathBuffer: LPSTR,
+    pcchBufferSize: LPDWORD, dwFlags: DWORD
+) -> PDH_STATUS}
+EXTERN!{stdcall fn PdhParseCounterPathW(
+    pCounterPathElements: *mut PDH_COUNTER_PATH_ELEMENTS_W, szFullPathBuffer: LPWSTR,
+    pcchBufferSize: LPDWORD, dwFlags: DWORD
+) -> PDH_STATUS}
+EXTERN!{stdcall fn PdhParseCounterPathA(
+    pCounterPathElements: *mut PDH_COUNTER_PATH_ELEMENTS_A, szFullPathBuffer: LPSTR,
```
Comment (yupi2) [inline src/um/pdh.rs]: The first two arguments to `PdhParseCounterPathA` and `PdhParseCounterPathW` [seem to be switched](https://msdn.microsoft.com/en-us/library/windows/desktop/aa372657(v=vs.85).aspx)

File: src/um/pdh.rs [hunk 6]
```diff
@@ -54,3 +222,338 @@ STRUCT!{struct PDH_COUNTER_PATH_ELEMENTS_W {
...
+) -> PDH_STATUS}
+EXTERN!{stdcall fn PdhOpenLogA(
+    szLogFileName: LPCSTR, dwAccessFlags: DWORD, lpdwLogType: LPDWORD, hQuery: PDH_HQUERY,
+    dwMaxSize: DWORD, szUserCaption: LPCSTR, phLog: *mut PDH_HLOG
+) -> PDH_STATUS}
+EXTERN!{stdcall fn PdhUpdateLogW(
+    hLog: PDH_HLOG, szUserString: LPCSTR
+) -> PDH_STATUS}
+EXTERN!{stdcall fn PdhUpdateLogA(
+    hLog: PDH_HLOG, szUserString: LPCWSTR
```
Comment (yupi2) [inline src/um/pdh.rs]: Swapped string types in `PdhUpdateLogW` and `PdhUpdateLogA`.

File: src/um/pdh.rs [hunk 7]
```diff
@@ -54,3 +222,338 @@ STRUCT!{struct PDH_COUNTER_PATH_ELEMENTS_W {
...
+) -> PDH_STATUS}
+EXTERN!{stdcall fn PdhGetDefaultPerfCounterHA(
+    hDataSource: PDH_HLOG, szMachineName: LPCSTR, szObjectName: LPCSTR,
+    szDefaultCounterName: LPSTR, pcchBufferSize: LPDWORD
+) -> PDH_STATUS}
+EXTERN!{stdcall fn PdhBrowseCountersHW(
+    pBrowseDlgData: PPDH_BROWSE_DLG_CONFIG_W
+) -> PDH_STATUS}
+EXTERN!{stdcall fn PdhBrowseCountersHA(
+    pBrowseDlgData: PPDH_BROWSE_DLG_CONFIG_A
```
Comment (yupi2) [inline src/um/pdh.rs]: `PdhBrowseCountersHA` and `PdhBrowseCountersHW` seem to take [PPDH_BROWSE_DLG_CONFIG_H*](https://msdn.microsoft.com/en-us/library/windows/desktop/aa373034(v=vs.85).aspx) not PPDH_BROWSE_DLG_CONFIG_*

**Comment by GuillaumeGomez** [comment]
@yupi2: Wo, excellent catches! Thanks!

**Comment by retep998** [comment]
There's actually a macro you can use to handle bitflags. You should consider using it.

**Comment by GuillaumeGomez** [comment]
Done as well.

File: src/um/winnt.rs [hunk 8]
```diff
@@ -38,6 +38,7 @@ pub type LPCWCH = *const WCHAR;
 pub type PCWCH = *const WCHAR;
 pub type NWPSTR = *mut WCHAR;
 pub type LPWSTR = *mut WCHAR;
+pub type LPTSTR = LPSTR;
```
Comment (retep998) [inline src/um/winnt.rs]: winapi doesn't provide the T variants of string types.

Comment (GuillaumeGomez) [inline src/um/winnt.rs]: Oh right, forgot to remove it.

**Comment by GuillaumeGomez** [comment]
Updated.

File: src/um/pdh.rs [hunk 9]
```diff
@@ -1,9 +1,14 @@
...
-use shared::minwindef::{DWORD};
-use shared::wtypesbase::{DOUBLE};
-use um::winnt::{LPWSTR, LPCWSTR, LPSTR, LPCSTR, LONGLONG, LONG, HANDLE};
+
+use shared::basetsd::DWORD_PTR;
+use shared::guiddef::GUID;
+use shared::minwindef::{BOOL, DWORD, FILETIME, LPDWORD, UCHAR};
+use shared::windef::HWND;
+use shared::wtypesbase::DOUBLE;
+use um::winnt::{LPWSTR, LPCWSTR, LPSTR, LPCSTR, LONGLONG, LONG, HANDLE, BOOLEAN};
```
Comment (retep998) [inline src/um/pdh.rs]: Don't forget to update `build.rs` to reflect the changes to header dependencies.

**Pull request merged**